streaming/base/stream.py

```diff
@@ -0,0 +1,312 @@
+                 download_timeout: Optional[float] = None,
+                 validate_hash: Optional[str] = None,
+                 keep_zip: Optional[bool] = None,
+                 keep_raw: Optional[bool] = None) -> None:
+        if not (remote or local):
+            raise ValueError('Remote and/or local path must be provided')
+
+        self.remote = remote
+
+        self._local = local
+        self.local = local or mkdtemp()
+
+        self.split = split or ''
+
+        if proportion is not None and repeat is not None:
+            raise ValueError('Proportion and repeat must not both be set -- one is derived from ' +
+                             'the other in relation to other Streams')
+
+        self._proportion = proportion
+        if proportion is not None:
+            if proportion <= 0:
+                raise ValueError('Proportion must be positive')
+            self.proportion = proportion
+
+        self._repeat = repeat
+        if repeat is not None:
+            if repeat <= 0:
+                raise ValueError('Repeat must be positive')
+            self.repeat = repeat
+
+        self._samples = samples
+        if samples is not None:
+            if samples <= 0:
+                raise ValueError('Samples must be positive')
+            self.samples = samples
+
+        self._download_retry = download_retry
+        if download_retry is not None:
+            if download_retry < 0:
+                raise ValueError('Download retry must be non-negative')
+            self.download_retry = download_retry
+
+        self._download_timeout = download_timeout
+        if download_timeout is not None:
+            if download_timeout <= 0:
+                raise ValueError('Download timeout must be positive')
+            self.download_timeout = download_timeout
+
+        self.validate_hash = validate_hash
+
+        self._keep_zip = keep_zip
+        if keep_zip is not None:
+            self.keep_zip = keep_zip
+
+        self._keep_raw = keep_raw
+        if keep_raw is not None:
+            self.keep_raw = keep_raw
+
+    def apply_default(self, default: Self) -> None:
+        """Apply defaults, setting any unset fields.
+
+        We use pairs of (name, _name) in order to make type checking happy.
+
+        Args:
+            default (Self): Stream containing default values for all optional fields.
+        """
+        if not self.split:
+            self.split = default.split or ''
+        if self._download_retry is None:
+            self.download_retry = default.download_retry
+        if self._download_timeout is None:
+            self.download_timeout = default.download_timeout
+        if self.validate_hash is None:
+            self.validate_hash = default.validate_hash or None
+        if self._keep_zip is None:
+            self.keep_zip = default.keep_zip
+        if self._keep_raw is None:
+            self.keep_raw = default.keep_raw
+
+    def _download_file(self, basename: str) -> str:
```

Comment:
are all these functions simply copied from where they used to live?

Reply:
yep

I moved the lower-level downloading/validating/decompressing out to the streams, because it depends on various stream-specific arguments. thankful to be reducing the size of the core StreamingDataset

I am not sure about where the shard_states logic (the layer above that) should ultimately live -- hopefully this will shake out well when adding dropping soon-to-be cold shards